### PR TITLE
Euro 2024: Using Fastly Compute

### PR DESCRIPTION
**Pull request by pmartin**
How we used Fastly Compute for Euro 2024 ;-)

### CODE REVIEW

**Review by shavounet — APPROVED** [review]
:ok_hand:

File: _posts/2024-11-07-compute-at-edge-personalize-static-pages.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+* At first, our KVStores’ primary location was in the US. Fastly reconfigured our account to have them primarily located in the EU, gaining several dozen milliseconds of latency each time new users would do their first read.
+* We compiled our Golang code to WASM with both Tinygo and Biggo. One produces WASM that used more CPU, and the other produces WASM that used more RAM. In the end, we followed Fastly’s recommendations, considering they know better than us what resource could be a bottleneck for their platform.
+* The first time we ran tests in our production environment, results were not great. Far worse than in our staging environment, in fact. Well, we had not paid for the Compute option in our Fastly's production account yet, and it was configured with lower rate-limits than in staging 😃
+
+During these load-tests as well as during real events later, we monitored a few basic metrics: number of calls per second, CPU and RAM usage, latency and error-rate.
+
+![](/images/posts/2024-11-compute-at-edge/compute-at-edge-monitoring-fastly-censored.png)
```
Comment (shavounet) [inline _posts/2024-11-07-compute-at-edge-personalize-static-pages.md:238]: Not fully censored !

**Review by shavounet — CHANGES_REQUESTED** [review]

File: _posts/2024-11-07-compute-at-edge-personalize-static-pages.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    api5[API-5]
+</div>
+
+<center><ins><strong>Schema 2: doing work in front of the BFF.</strong></ins></center>
+
+Our goal was to implement a lightweight personalization layer in front of our backend BFF application. The backend application would then only need to return non-personalized layouts -- and those would be cacheable.
+
+Note: as we have many frontend applications and did not want (and did not have enough time) to update all of them, we searched for a solution that would only require changes somewhere on the backend side.
+
+This personalization layer would read some basic data *(does the current user have a subscription? Did they consent to tracking and analytics?)* from a datastore-at-edge and use them to inject identifiers in the cached-non-personalized layout before returning it.
```
Comment (fragab) [inline _posts/2024-11-07-compute-at-edge-personalize-static-pages.md:113]: It mixes singular and plural, what about `Did he consent to tracking and analytics?`

**Review by fragab — APPROVED** [review]
Great article 👏

File: _posts/2024-11-07-compute-at-edge-personalize-static-pages.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    api3[API-3]
+    api4[API-4]
+    api5[API-5]
+</div>
+
+<center><ins><strong>Schema 1: one BFF between frontend apps and backend APIs.</strong></ins></center>
+
+With this architecture, BFF receives a lot of requests, and each of these causes up to dozens of requests to other services. This is greatly amplified by the fact that all BFF responses are personalized, which means we cannot cache entire responses.
+
+As most of our APIs are running in a Kubernetes cluster, you might say *“use auto-scaling!”*, and you’d be kind of right. We are using auto-scaling. But reactive auto-scaling, based on traffic or load metrics, is not fast enough to handle huge spikes in traffic -- like big football matches can cause. For this, we have developed a [pre-scaling mechanism](/2022/02/03/prescaling.html), and even [open-sourced it](tech.bedrockstreaming.com).
```
Comment (fdubost) [inline _posts/2024-11-07-compute-at-edge-personalize-static-pages.md:52]: The open source lin seems not good

File: _posts/2024-11-07-compute-at-edge-personalize-static-pages.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    bff --> api5
+
+    api1[API-1]
+    api2[API-2]
+    api3[API-3]
+    api4[API-4]
+    api5[API-5]
+</div>
+
+<center><ins><strong>Schema 1: one BFF between frontend apps and backend APIs.</strong></ins></center>
```
Comment (fdubost) [inline _posts/2024-11-07-compute-at-edge-personalize-static-pages.md:48]: Don't know why, but on the preview the legend is stuck to the following paragraph 🤷

**Review by fdubost — APPROVED** [review]
Very clear 👏 
(said by a frontend dev 😄)

File: _posts/2024-11-07-compute-at-edge-personalize-static-pages.md (hunk 4 above)
Comment (pmartin) [reply _posts/2024-11-07-compute-at-edge-personalize-static-pages.md:48]: 😭 I tried a few things and ended-up adding a `<br>` after each legend. Yeah, ugly… But works 🤷‍♂️ (and I don't dare touching CSS)

**Review by shavounet — APPROVED** [review]

**Pull request merged**